biz.aQute.bndlib/src/aQute/bnd/build/Workspace.java

```diff
@@ -214,26 +214,35 @@ public static Workspace getWorkspace(File parent, String bndDir) throws Exceptio
 		}
 	}
 
-	public Workspace(File dir) throws Exception {
-		this(dir, CNFDIR);
+	public Workspace(File workspaceDir) throws Exception {
+		this(workspaceDir, CNFDIR);
 	}
 
-	public Workspace(File dir, String bndDir) throws Exception {
+	public Workspace(File workspaceDir, String bndDir) throws Exception {
 		super(getDefaults());
+		workspaceDir = workspaceDir.getAbsoluteFile();
+		setBase(workspaceDir); // setBase before call to setFileSystem
 		this.layout = WorkspaceLayout.BND;
 		addBasicPlugin(new LoggingProgressPlugin());
-		setFileSystem(dir, bndDir);
+		setFileSystem(workspaceDir, bndDir);
 	}
 
-	public void setFileSystem(File workspaceDir, String nameOfCnf) throws IOException {
-
+	public void setFileSystem(File workspaceDir, String bndDir) throws Exception {
 		workspaceDir = workspaceDir.getAbsoluteFile();
 		if (!workspaceDir.exists() && !workspaceDir.mkdirs()) {
 			throw new IOException("Could not create directory " + workspaceDir);
 		}
 		assert workspaceDir.isDirectory();
 
-		File buildDir = new File(workspaceDir, nameOfCnf).getAbsoluteFile();
+		synchronized (cache) {
+			WeakReference<Workspace> wsr = cache.get(getBase());
+			if ((wsr != null) && (wsr.get() == this)) {
```

Comment:
We need to do the get since there can be multiple Workspace objects (via new) for a given dir. But only one can be in the cache, so we need to confirm we are that one before removing from the cache.